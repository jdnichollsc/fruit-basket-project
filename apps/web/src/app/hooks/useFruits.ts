import { useState, useEffect, useCallback } from 'react';
import type { FruitState } from '../types/fruit';

export function useFruits() {
  const [state, setState] = useState<FruitState>({
    fruits: [],
    loading: true,
    error: null,
  });
  const [isSubmitting, setIsSubmitting] = useState(false);

  // Reusable error handler for API operations
  const handleApiOperation = async <T>(
    operation: () => Promise<T>,
    errorMessage: string,
    options?: { setSubmitting?: boolean }
  ) => {
    if (options?.setSubmitting) {
      setIsSubmitting(true);
    }
    setState((prev) => ({ ...prev, error: null }));

    try {
      const result = await operation();
      return result;
    } catch (error) {
      setState((prev) => ({
        ...prev,
        error: error instanceof Error ? error.message : errorMessage,
      }));
      return false;
    } finally {
      if (options?.setSubmitting) {
        setIsSubmitting(false);
      }
    }
  };

  const loadFruits = useCallback(async () => {
    setState((prev) => ({ ...prev, loading: true, error: null }));
    const fruits = await handleApiOperation(
      async () => window.BASKET.API.getAll(),
      'Failed to load fruits'
    );
    setState((prev) => ({
      ...prev,
      fruits: Array.isArray(fruits) ? fruits : prev.fruits,
      loading: false,
    }));
  }, []);

  useEffect(() => {
    loadFruits();
  }, [loadFruits]);

  const addFruit = async (name: string) => {
    const fruits = await handleApiOperation(
      async () => window.BASKET.API.add(name),
      'Failed to add fruit',
      { setSubmitting: true }
    );
    if (Array.isArray(fruits)) {
      setState((prev) => ({ ...prev, fruits }));
      return true;
    }
    return false;
  };

  const updateFruit = async (oldName: string, newName: string) => {
    const fruits = await handleApiOperation(
      async () => window.BASKET.API.update(oldName, newName),
      'Failed to update fruit'
    );
    if (Array.isArray(fruits)) {
      setState((prev) => ({ ...prev, fruits }));
      return true;
    }
    return false;
  };

  const deleteFruit = async (name: string) => {
    const fruits = await handleApiOperation(
      async () => window.BASKET.API.delete(name),
      'Failed to delete fruit'
    );
    if (Array.isArray(fruits)) {
      setState((prev) => ({ ...prev, fruits }));
      return true;
    }
    return false;
  };

  return {
    ...state,
    isSubmitting,
    addFruit,
    updateFruit,
    deleteFruit,
  };
}
