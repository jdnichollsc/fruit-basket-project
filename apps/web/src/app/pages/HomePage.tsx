import React from 'react';
import { FruitForm, FruitList } from '@fruit-basket/ui';
import { useFruits } from '../hooks/useFruits';

export function HomePage() {
  const {
    fruits,
    loading,
    error,
    isSubmitting,
    addFruit,
    updateFruit,
    deleteFruit,
  } = useFruits();

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div className="text-red-500">{error}</div>;
  }

  return (
    <div className="container mx-auto p-4">
      <h1 className="text-2xl font-bold mb-4">
        Fruit Basket{' '}
        <span role="img" aria-label="shopping cart">🛒</span>
        <span role="img" aria-label="chart increasing">📈</span>
        <span role="img" aria-label="tomato">🍅</span>
        <span role="img" aria-label="shopping bags">🛍️</span>
      </h1>
      <FruitForm onAdd={addFruit} isLoading={isSubmitting} error={error} />
      <FruitList
        fruits={fruits}
        isLoading={isSubmitting}
        onEdit={async (...rest) => {
          updateFruit(...rest)
        }}
        onDelete={async (...rest) => {
          await deleteFruit(...rest)
        }}
      />
    </div>
  );
}
