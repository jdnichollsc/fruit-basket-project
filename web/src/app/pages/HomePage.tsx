import { FruitForm, FruitList } from '@fruit-basket/ui';
import { useFruits } from '../hooks/useFruits';
import { promptForNewName } from '../services/prompt';

export function HomePage() {
  const { fruits, loading, error, isSubmitting, addFruit, updateFruit, deleteFruit } = useFruits();
  
  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div className="text-red-500">{error}</div>;
  }

  if (isSubmitting) {
    return <div>Submitting...</div>;
  }

  return (
    <div className="container mx-auto p-4">
      <h1 className="text-2xl font-bold mb-4">Fruit Basket</h1>
      <FruitForm onAdd={addFruit} isLoading={isSubmitting} error={error} />
      <FruitList
        fruits={fruits}
        onEdit={(fruit: string) => {
          const newName = promptForNewName(fruit);
          if (newName) {
            updateFruit(fruit, newName);
          }
        }}
        onDelete={deleteFruit}
      />
    </div>
  );
} 