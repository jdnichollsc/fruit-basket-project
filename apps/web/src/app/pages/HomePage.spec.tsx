import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, within } from '@testing-library/react';
import { HomePage } from './HomePage';
import { useFruits } from '../hooks/useFruits';
import type { FruitFormProps, FruitListProps } from '@fruit-basket/ui';

// Mock the dependencies
jest.mock('../hooks/useFruits');
jest.mock('@fruit-basket/ui', () => ({
  FruitForm: ({ onAdd, isLoading, error }: Partial<FruitFormProps>) => (
    <div data-testid="fruit-form" onClick={() => onAdd?.('Cherry')}>
      Fruit Form {isLoading ? 'Loading' : ''} {error || ''}
    </div>
  ),
  FruitList: ({ fruits, isLoading, onEdit, onDelete }: Partial<FruitListProps>) => (
    <div data-testid="fruit-list">
      Fruit List {isLoading ? 'Loading' : ''}
      {fruits?.map((fruit: string) => (
        <div key={fruit} data-testid={`fruit-item-${fruit}`}>
          <span>{fruit}</span>
          <button onClick={() => onEdit?.(fruit, 'Updated ' + fruit)}>Edit</button>
          <button onClick={() => onDelete?.(fruit)}>Delete</button>
        </div>
      ))}
    </div>
  ),
}));

describe('HomePage', () => {
  const mockUseFruits = useFruits as jest.Mock;
  const mockAddFruit = jest.fn().mockResolvedValue(true);
  const mockUpdateFruit = jest.fn().mockResolvedValue(true);
  const mockDeleteFruit = jest.fn().mockResolvedValue(true);

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseFruits.mockReturnValue({
      fruits: ['Apple', 'Banana'],
      loading: false,
      error: null,
      isSubmitting: false,
      addFruit: mockAddFruit,
      updateFruit: mockUpdateFruit,
      deleteFruit: mockDeleteFruit,
    });
  });

  it('should render loading state', () => {
    mockUseFruits.mockReturnValue({
      fruits: [],
      loading: true,
      error: null,
      isSubmitting: false,
      addFruit: jest.fn(),
      updateFruit: jest.fn(),
      deleteFruit: jest.fn(),
    });

    render(<HomePage />);
    expect(screen.getByText('Loading Fruit Basket...')).toBeInTheDocument();
    expect(screen.getAllByRole('generic').filter(el => el.className.includes('skeleton'))).not.toHaveLength(0);
  });

  it('should render error state', () => {
    const errorMessage = 'Failed to load fruits';
    mockUseFruits.mockReturnValue({
      fruits: [],
      loading: false,
      error: errorMessage,
      isSubmitting: false,
      addFruit: jest.fn(),
      updateFruit: jest.fn(),
      deleteFruit: jest.fn(),
    });

    render(<HomePage />);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it('should render fruit form and list when data is loaded', () => {
    render(<HomePage />);
    expect(screen.getByTestId('fruit-form')).toBeInTheDocument();
    expect(screen.getByTestId('fruit-list')).toBeInTheDocument();
  });

  it('should handle adding a fruit', async () => {
    render(<HomePage />);
    const fruitForm = screen.getByTestId('fruit-form');
    await fireEvent.click(fruitForm);
    expect(mockAddFruit).toHaveBeenCalledWith('Cherry');
  });

  it('should handle editing a fruit', async () => {
    render(<HomePage />);
    const appleItem = screen.getByTestId('fruit-item-Apple');
    const editButton = within(appleItem).getByText('Edit');
    await fireEvent.click(editButton);
    expect(mockUpdateFruit).toHaveBeenCalledWith('Apple', 'Updated Apple');
  });

  it('should handle deleting a fruit', async () => {
    render(<HomePage />);
    const appleItem = screen.getByTestId('fruit-item-Apple');
    const deleteButton = within(appleItem).getByText('Delete');
    await fireEvent.click(deleteButton);
    expect(mockDeleteFruit).toHaveBeenCalledWith('Apple');
  });

  it('should show loading state in form and list during submission', () => {
    mockUseFruits.mockReturnValue({
      fruits: ['Apple', 'Banana'],
      loading: false,
      error: null,
      isSubmitting: true,
      addFruit: mockAddFruit,
      updateFruit: mockUpdateFruit,
      deleteFruit: mockDeleteFruit,
    });

    render(<HomePage />);
    expect(screen.getByTestId('fruit-form')).toHaveTextContent('Loading');
    expect(screen.getByTestId('fruit-list')).toHaveTextContent('Loading');
  });

  it('should match snapshot in loading state', () => {
    mockUseFruits.mockReturnValue({
      fruits: [],
      isLoading: true,
      error: null,
      addFruit: jest.fn(),
      editFruit: jest.fn(),
      deleteFruit: jest.fn(),
    });
    const { container } = render(<HomePage />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with fruits', () => {
    mockUseFruits.mockReturnValue({
      fruits: ['Apple', 'Banana'],
      isLoading: false,
      error: null,
      addFruit: jest.fn(),
      editFruit: jest.fn(),
      deleteFruit: jest.fn(),
    });
    const { container } = render(<HomePage />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with error', () => {
    mockUseFruits.mockReturnValue({
      fruits: [],
      isLoading: false,
      error: 'Failed to load fruits',
      addFruit: jest.fn(),
      editFruit: jest.fn(),
      deleteFruit: jest.fn(),
    });
    const { container } = render(<HomePage />);
    expect(container).toMatchSnapshot();
  });
});
