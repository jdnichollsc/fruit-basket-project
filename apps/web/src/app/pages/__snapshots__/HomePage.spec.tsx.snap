// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match snapshot in loading state 1`] = `
<div>
  <div
    class="container mx-auto p-4"
  >
    <h1
      class="text-2xl font-bold mb-4"
      data-testid="page-title"
    >
      Fruit Basket
       
      <span
        aria-label="shopping cart"
        role="img"
      >
        🛒
      </span>
      <span
        aria-label="tomato"
        role="img"
      >
        🍅
      </span>
      <span
        aria-label="shopping bags"
        role="img"
      >
        🛍️
      </span>
    </h1>
    <div
      data-testid="fruit-form"
    >
      Fruit Form 
       
    </div>
    <div
      data-testid="fruit-list"
    >
      Fruit List 
    </div>
  </div>
</div>
`;

exports[`HomePage should match snapshot with error 1`] = `
<div>
  <div
    class="text-red-500"
    data-testid="error-message"
  >
    Failed to load fruits
  </div>
</div>
`;

exports[`HomePage should match snapshot with fruits 1`] = `
<div>
  <div
    class="container mx-auto p-4"
  >
    <h1
      class="text-2xl font-bold mb-4"
      data-testid="page-title"
    >
      Fruit Basket
       
      <span
        aria-label="shopping cart"
        role="img"
      >
        🛒
      </span>
      <span
        aria-label="tomato"
        role="img"
      >
        🍅
      </span>
      <span
        aria-label="shopping bags"
        role="img"
      >
        🛍️
      </span>
    </h1>
    <div
      data-testid="fruit-form"
    >
      Fruit Form 
       
    </div>
    <div
      data-testid="fruit-list"
    >
      Fruit List 
      <div
        data-testid="fruit-item-Apple"
      >
        <span>
          Apple
        </span>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div
        data-testid="fruit-item-Banana"
      >
        <span>
          Banana
        </span>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
